third_party/amd/lib/TritonAMDGPUTransforms/ReorderInstructions.cpp

```diff
@@ -97,6 +98,51 @@ class TritonAMDGPUReorderInstructionsPass
       kv.first->moveBefore(kv.second);
     opToMove.clear();
 
+    // Move writing to LDS and reading from LDS right after the loading of a
+    // tensor from global memory. There are 2 possible patterns depending on
+    // whether writing to LDS is done using an optional local_alloc argument or
+    // a local_store instruction:
+    //
+    // 1) %1 = load %ptr
+    //    %2 = local_alloc %1
+    //    %3 = local_load %2
+    //
+    // 2) %1 = load %ptr
+    //    %2 = local_alloc
+    //    %3 = local_store %1, %2
+    //    %4 = local_load %2
+    m.walk([&](ttg::LocalLoadOp op) {
+      Operation *localLoad = (Operation *)op;
+      Operation *localAlloc = localLoad->getOperand(0).getDefiningOp();
+
+      if (!localAlloc || !isa<ttg::LocalAllocOp>(localAlloc)) {
+        return;
+      }
+
+      // Case when localAlloc has no operands
+      if (localAlloc->getNumOperands() < 1) {
+        Operation *localStore = getFirstUse(localAlloc);
+        assert(localStore && "localAlloc has no use");
+
+        // Check if the use is a LocalStoreOp
+        if (isa<ttg::LocalStoreOp>(localStore)) {
+          Operation *loadOp = localStore->getOperand(0).getDefiningOp();
+
+          if (isa<tt::LoadOp>(loadOp)) {
+            localAlloc->moveAfter(loadOp);
```

Comment:
What about all the pointer logic that these ops depend on? 
You must capture/move the cone of logic for all operands of the localStore. 
The `local_store` may also be in the loop and it may not be valid to hoist it.

Reply:
Check localAlloc->getBlock() == localStore->getBlock() == loadOp->getBlock() to verify in the same block.

And check below where the logic cone is captured (line 130).

Reply:
+1